### PR TITLE
Added a new route for supporting a secondary bucket

### PR DESCRIPTION
**Pull request by som-poddar**
- Ad crunchyroll, we would like to extend this utility to support avod streaming in the platform.
- This change adds this capability with the minimal amount of change
to the existing functionality.
- Updated to use aws sdk for s3 get call

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -50,8 +60,8 @@ but this can be changed with the -config option, e.g. "-config=./test.yml"
     s3_path:    <optional prefix to prepend to object requests>
     s3_retries; <maximum number of S3 retries>
     s3_timeout: <timeout for S3 requests>
-    
-    
+```
```
Comment (rmikhael) [inline README.md:63]: add s3_ad_bucket to the config params example.

**Comment by mskariah1942** [comment]
@som-poddar where are the unit tests?

**Review by rmikhael — APPROVED** [review]
+1

File: main.go [hunk 2]
```diff
@@ -63,7 +63,9 @@ func main() {
 	initRuntime()
 
 	mux := http.NewServeMux()
-	mux.Handle("/", http.HandlerFunc(forwardToS3))
+
+	mux.Handle("/avod/", http.HandlerFunc(forwardToS3ForAd))
```
Comment (mlevkov) [inline main.go:67]: @som-poddar why are we sticking to the single bucket, it would be better if the route was reflecting a bucket name instead of this "fixed" route. as such, /{bucket Name}/<key.... etc.>

File: s3.go [hunk 3]
```diff
@@ -41,7 +41,52 @@ func initRuntime() {
...
+	w.Header().Set("Server", serverName)
+
+	if r.Method != "GET" && r.Method != "HEAD" {
+		w.WriteHeader(405)
+		return
+	}
+
+	// Make sure that Remote Address is 127.0.0.1 so it comes off a local proxy
+	a := strings.SplitN(r.RemoteAddr, ":", 2)
+	if len(a) != 2 || a[0] != "127.0.0.1" {
```
Comment (mlevkov) [inline s3.go:54]: At some point, this needs to change. There is no reason why it has to be "local". Nginx has no such restriction.

File: s3.go [hunk 4]
```diff
@@ -63,7 +108,7 @@ func forwardToS3(w http.ResponseWriter, r *http.Request) {
 		Str("range", byterange).
 		Str("method", r.Method).
 		Logger()
-	s3url := fmt.Sprintf("http://s3-%s.amazonaws.com/%s%s%s", conf.S3Region, conf.S3Bucket, conf.S3Path, upath)
+	s3url := fmt.Sprintf("http://s3-%s.amazonaws.com/%s%s%s", conf.S3Region, bucket, conf.S3Path, upath)
```
Comment (mlevkov) [inline s3.go:111]: This is a very expensive operation on every call, try to consider using Join.

File: s3client/get.go [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+	getInput := &s3.GetObjectInput{
+		Bucket: aws.String(bucket),
+		Key:    aws.String(s3Path),
+		Range:  aws.String(byterange),
+	}
+
+	sess, err := session.NewSessionWithOptions(session.Options{
+		Config:            aws.Config{Region: aws.String("us-west-2")},
+		SharedConfigState: session.SharedConfigEnable,
+	})
```
Comment (mlevkov) [inline s3client/get.go:20]: why are you making a new session on every call to this function?

File: s3.go [hunk 6]
```diff
@@ -41,7 +41,52 @@ func initRuntime() {
...
+		return
+	}
+
+	p := []rune(r.URL.Path)
+	s3Path := string(p[6:])
+	byterange := r.Header.Get("Range")
+	logger := log.With().
+		Str("object", s3Path).Str("range", byterange).Str("method", r.Method).Logger()
+
+	getObject, getErr := s3client.GetObject(conf.S3AdBucket, s3Path, byterange)
```
Comment (mlevkov) [inline s3.go:65]: calling GetObject with a new session set up on every request might get very expensive and blocking.

**Review by mlevkov — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (som-poddar) [reply README.md:64]: done.

File: s3.go (hunk 4 above)
Comment (som-poddar) [reply s3.go:110]: Agreed. again, the intent of this PR is to keep the current plumbing for the media route exactly the same while adding support for the ad media route which will use a secondary bucket.

I will address move all performance optimizations to a different PR via new Github issues.

File: s3.go (hunk 6 above)
Comment (som-poddar) [reply s3.go]: Fair point @mlevkov, I was aware of this limitation but was reluctant to add this in one PR to keep the diff small/manageable. But on second thought, made the changes you suggested via https://github.com/crunchyroll/evs-s3helper/pull/27/commits/ec124ef838e62db0a524285775cf2cfc0c4a08f3.

File: s3client/get.go (hunk 5 above)
Comment (som-poddar) [reply s3client/get.go]: see my response https://github.com/crunchyroll/evs-s3helper/pull/27#discussion_r676938467

File: s3.go (hunk 3 above)
Comment (som-poddar) [reply s3.go]: done. Filed https://github.com/crunchyroll/evs-s3helper/issues/29

File: main.go (hunk 2 above)
Comment (som-poddar) [reply main.go]: @mlevkov, we are not sticking to a single bucket, matter of fact - the purpose of this PR is to introduce a secondary bucket to serve AVOD contents. 

The reason this service exists is to abstract downstream bucket configuration, permissions, etc from the caller. For example, the VoD Nginx packager only needs to know, which route leads to serving ad content and which route leads to serving VoD content.  Anything beyond the route must be abstracted from the packager or upstream caller(s).

Comment (mlevkov) [reply main.go]: totally understood, but this design is only specific to the "avod" bucket. The consideration should be defined via bucket abstract, which is accessible from Nginx config. If you consider the design I've proposed, then you can easily expose "other" buckets without rewriting the software in the future, whenever we need new buckets to access. This is too much coupling to bucket name and purpose.

File: s3.go (hunk 4 above)
Comment (mlevkov) [reply s3.go:110]: changing Sprintf to join is not that difficult. How do you know that your "sprintf" call will not create an additional burden on already heavy call-oriented adapter? Have you profiled your code?

File: s3client/get.go (hunk 5 above)
Comment (mlevkov) [reply s3client/get.go]: I saw that, thanks. Excellent!

**Comment by mlevkov** [comment]
please consider adding unit tests, positive and negative, even if it is a mocked payload. Please test this thoroughly in Proto0.

**Comment by mskariah1942** [comment]
Since this PR is now more than 20 days, are we still waiting to get unit tests in? Let me know if there are concerns or any discussion needed. Thanks!

File: s3.go [hunk 7]
```diff
@@ -41,7 +40,50 @@ func initRuntime() {
...
+
+	io.Copy(w, getObject.Body)
+	logger.Info().Str("path", s3Path).Int64("content-length", *getObject.ContentLength).Msg("s3:get - success")
+}
+
+func forwardToS3ForMedia(w http.ResponseWriter, r *http.Request) {
+	forwardToS3(w, r, conf.S3Bucket)
+}
+
+func forwardToS3(w http.ResponseWriter, r *http.Request, bucket string) {
```
Comment (rmikhael) [inline s3.go:86]: I know you didn't change that, but shouldn't all 3 "forwardToS3" functions  be methods of App, for consistency?

Comment (som-poddar) [reply s3.go:86]: Noted, perhaps a new issue/pull-req will be a good place for this.
